File: golang/aoc/day01.go
package aoc

import (
	"fmt"
)

func Day01(day int) {
	fmt.Printf("...1: %s\n", challenge1(day))
	fmt.Printf("...2: %s\n", challenge2(day))
}

func challenge1(day int) string {
	integers, err := AsInts(day, 1)
	if err != nil {
		panic(err)
	}
	increases := countDepthIncreasesWin(integers, 2)
	return fmt.Sprintf("Increasing depth: %d times", increases)
}

func challenge2(day int) string {
	integers, err := AsInts(day, 1)
	if err != nil {
		panic(err)
	}
	increases := countDepthIncreasesWin(integers, 4)
	return fmt.Sprintf("Increasing depth in windows of 3: %d times", increases)
}

func countDepthIncreasesWin(measurements []int64, window int) int {
	count := 0
	for i := 0; i < len(measurements)-window+1; i++ {
		if measurements[i] < measurements[i+window-1] {
			count++
		}
	}
	return count
}

File: rust/src/aoc/day01.rs
use crate::aoc::input;

pub fn main(day: usize) {
    println!("...1: {}", challenge1(day));
    println!("...2: {}", challenge2(day));
}

pub fn challenge1(day: usize) -> String {
    let integers = input::as_ints(day, 1);
    if let Ok(integers) = integers {
        let increases = count_depth_increases_win(&integers, 2);
        return format!("Increasing depth: {} times", increases);
    }
    "".to_string()
}

pub fn challenge2(day: usize) -> String {
    let integers = input::as_ints(day, 1);
    if let Ok(integers) = integers {
        let increases = count_depth_increases_win(&integers, 4);
        return format!("Increasing depth in windows of 3: {} times", increases);
    }
    "".to_string()
}

fn count_depth_increases_win(measurements: &Vec<i64>, window: usize) -> usize {
    measurements
        .windows(window)
        .filter_map(|w| match w.first().copied() < w.last().copied() {
            true => Some(true),
            false => None,
        })
        .count()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn challenge1() {
        let numbers = vec![199, 200, 208, 210, 200, 207, 240, 269, 260, 263];
        let count = count_depth_increases_win(&numbers, 2);
        assert_eq!(count, 7)
    }

    #[test]
    fn challenge2() {
        let numbers = vec![199, 200, 208, 210, 200, 207, 240, 269, 260, 263];
        let count = count_depth_increases_win(&numbers, 4);
        assert_eq!(count, 5);
    }
}
